### PR TITLE
Dev container 

### PR DESCRIPTION
**Pull request by bHimes**
# Description

This will need some polishing, but is setup to be a fully contained build environment, tested on a clean ubuntu and even wsl2 on my windows box.

- pulled from Ubuntu 20.04 image
- gcc-11 (gets most complete c++20 set, but only c++17 tested)
- wxWidgets 3.05
- wxFormbuilder 3.10
  - Runs but did needs to be tested to make sure it plays nice with wx 3.05, ideally, test 
  - wx 3.1 with gtk3
- clang-format-14
- cuda 11.6
  - I deleted several libraries we don't use to get this from ~6.4 -> ~1.9 Gb
- libtorch
  - This adds ~3.8 Gb (total image is ~5.8 compressed) but I plan to start testing the c++ API for pytorch directly. Could compile and run a simple test program (outside cisTEMx, but in the build env)

# I have rebased my feature branch to be current with the master branch using to minimize conflicts and headaches

- [X] yes
- [ ] no

# Which compilers were tested

- [X] g++
- [ ] icpc
- [ ] clang
- [ ] other (please specify)

# These changes are isolated to the

- [ ] gui
- [ ] core library
- [ ] gpu core library
- [X] program it modifies

# How has the functionality been tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

Ran builds on several computers, including a naive bare metal ubuntu install (20) and a virtualized wsl2 (18)

Tested the build itself from outside the docker container, ran basic tests + template matching and seems to be good.

# TODO:

@jojoelfe I was thinking something like:
1 - test with g++20 for CPU components (cuda limited to 17)
2 - test with dynamic libs
  - This will further reduce the size of the libraries we need to include shrinking the container
  - May be an issue with cufft callbacks, but these will be replaced by FastFFT generally, and can just be skipped otherwise. 
  - Then for running, switch to a singularity (now apptainer) run which we should be able to translate directly from the docker image (or perhaps a slightly more lightweight version)
3 - Opt for submodules to directly include cisTEM_docs, FastFFT, and GEMMI
  - submodules will also allow us to test private features by having them in a separate repo that is a submodule and under control of a "DONOTDISTRIBUTE" flag (or something.) A particular use case I'm thinking is testing a direct port of cryoDrgon. Even though it is currently GPLv3, keeping it private and not releasing publically is okay. Then, if it works, Ellen could release a version with a non viral license so cisTEMx doesn't get infected. (I've spoken with her about this.)

### CODE REVIEW

File: .devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
```
Comment (jojoelfe) [inline .devcontainer.json:1]: Why is this file present twice? .devcontainer and .devcontainer_build/.devcontainer

File: .devcontainer_build/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    cd lib64 && rm libcublas* libcuspars* libcusolver*
+
+RUN add-apt-repository ppa:ubuntu-toolchain-r/test -y && \
+    apt-get --allow-releaseinfo-change  update && apt install -y gcc-${GCC_VER} g++-${GCC_VER} \
+    && update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-${GCC_VER} 100 --slave \
+    /usr/bin/g++ g++ /usr/bin/g++-${GCC_VER} --slave /usr/bin/gcov gcov /usr/bin/gcov-${GCC_VER} \
+    && rm -rf /var/lib/apt/lists/*
+
+#Need newer git for Github actions
+RUN add-apt-repository ppa:git-core/ppa -y && apt-get --allow-releaseinfo-change update && apt-get install -y git \
```
Comment (jojoelfe) [inline .devcontainer_build/Dockerfile:54]: This might not be required for ubuntu 20.0

File: .vscode_shared/BenHimes/tasks.json [hunk 3]
```diff
@@ -91,18 +86,23 @@
...
-            "command": "cd  ${build_dir}/GNU-gpu && make -j${compile_cores}"
+            "command": "cd  ${build_dir}/GNU-gpu && make -j4"
         },
-   
-
+        {
+            // This doesn't work - spaces, ugh
+            "label": "Build dev container",
+            "type": "shell",
+            "command": "~/.config/devcontainer build --image-name bhimesbhimes/cistem_build_env .devcontainer_build/"
```
Comment (jojoelfe) [inline .vscode_shared/BenHimes/tasks.json:103]: Why is the ~/.config/devontainer needed?

**Review by jojoelfe — APPROVED** [review]
Just a few question, to understand better whats going on.

File: .devcontainer.json (hunk 1 above)
Comment (bHimes) [reply .devcontainer.json:1]: They aren't quite the same:

.devcontainer.json
- no build: block, instead references an image name. This is used when doing something like Remote-Contianers: Open Folder in Container. So vscode easily opens the folder you want in the container it spins up from image at "image:"

.devcontainer_build/.devcontainer.json
- with build: block, which points to the Dockerfile. This is the recipe for the actual build.

File: .devcontainer_build/Dockerfile (hunk 2 above)
Comment (bHimes) [reply .devcontainer_build/Dockerfile:54]: Good point - I'll check it out (#47)

File: .vscode_shared/BenHimes/tasks.json (hunk 3 above)
Comment (bHimes) [reply .vscode_shared/BenHimes/tasks.json:103]: This isn't actually a devcontainer, it is a program provided by the Remote-Containers CLI tool. This is used from the CLI to build the container in a way that vscode is also happy with it. Unfortunately there is the stupid space in the path, so this is just here as a record but must be entered on the CLI with escapes "\ -\ "

**Pull request merged**